### PR TITLE
Add totalBytesSent metric to tcp_client

### PR DESCRIPTION
**Pull request by rahultejwani**
**What this PR does / why we need it**:

Adding metrics to keep track of total bytes sent from the tcp_client
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/aggregator/client/tcp_client.go [hunk 1]
```diff
@@ -301,7 +302,9 @@ func (c *TCPClient) write(
 			multiErr = multiErr.Add(err)
 			continue
 		}
-
+		// using unsfae.Sizeof to avoid the overhead of marshaling the payload to bytes
+		// tradeoff is that this is not 100% accurate, but should be close enough
+		c.metrics.totalBytesSent.Inc(int64(unsafe.Sizeof(payload)))
```
Comment (prateek) [inline src/aggregator/client/tcp_client.go:307]: using `unsafe.Sizeof` is going to be way off. 

> Sizeof takes an expression x of any type and returns the size in bytes of a hypothetical variable v as if v was declared via var v = x. **The size does not include any memory possibly referenced by x.** For instance, if x is a slice, Sizeof returns the size of the slice descriptor, not the size of the memory referenced by the slice; if x is an interface, Sizeof returns the size of the interface value itself, not the size of the value stored in the interface. For a struct, the size includes any padding introduced by field alignment. The return value of Sizeof is a Go constant if the type of the argument x does not have variable size. (A type has variable size if it is a type parameter or if it is an array or struct type with elements of variable size).

via [godocs](https://pkg.go.dev/unsafe#Sizeof)

Comment (prateek) [reply src/aggregator/client/tcp_client.go:307]: as part of writing it out (the `.Write()` call above), the code is already traversing the object to encode it 

it'll be a minor change to bubble up the difference in encoder length, after writing from this [method](https://github.com/m3db/m3/blob/master/src/aggregator/client/writer.go#L170). 

Please do that instead.

Comment (rahultejwani) [reply src/aggregator/client/tcp_client.go:307]: Ah great point. I will update

File: src/aggregator/client/writer.go [hunk 2]
```diff
@@ -93,39 +93,43 @@ func newInstanceWriter(instance placement.Instance, opts Options) instanceWriter
...
-		w.RUnlock()
-		return err
+		// If the encoder exists, encode without acquiring the write lock.
+		return w.encodeWithLock(encoder, payload)
 	}
-	w.RUnlock()
 
+	// Acquire the write lock if the encoder doesn't exist.
 	w.Lock()
+	defer w.Unlock()
```
Comment (prateek) [inline src/aggregator/client/writer.go:114]: `defer()` prevents inlining in all kinds of weird ways. can you avoid relying on that here.

File: src/aggregator/client/writer.go [hunk 3]
```diff
@@ -163,13 +171,12 @@ func (w *writer) QueueSize() int {
 func (w *writer) encodeWithLock(
 	encoder *lockedEncoder,
 	payload payloadUnion,
-) error {
+) (int, error) {
 	encoder.Lock()
+	defer encoder.Unlock()
```
Comment (prateek) [inline src/aggregator/client/writer.go:176]: same point about defer

File: src/aggregator/client/writer_mgr.go [hunk 4]
```diff
@@ -188,23 +188,24 @@ func (mgr *writerManager) Write(
...
-		mgr.RUnlock()
-		return fmt.Errorf("writer for instance %s is not found", id)
+		return 0, fmt.Errorf("writer for instance %s not found", id)
 	}
-	writer.dirty.Store(true)
-	err := writer.Write(shardID, payload)
-	mgr.RUnlock()
 
-	return err
+	writer.dirty.Store(true)
```
Comment (prateek) [inline src/aggregator/client/writer_mgr.go:207]: i don't know if the code relies on this semantic or not - but earlier we held the read lock when marking the dirty bit - we no longer do that after your change.

can you ensure that change is safe / if you're not sure - go back to the old way.

**Review by prateek — APPROVED** [review]
lgtm sans the dirty bit issue

File: src/aggregator/client/writer_mgr.go (hunk 4 above)
Comment (rahultejwani) [reply src/aggregator/client/writer_mgr.go:205]: it was outside of the scope or read lock, but I reverted that optimization to avoid any unknowns

**Pull request closed without merging**